### PR TITLE
[WIP] gas_mixture twixes

### PR DESCRIPTION
**Pull request by Dennok**
## About The Pull Request

garbage_collect() now remove all gases with mole amount smaller than MINIMUM_MOLE_COUNT = 0.01
This must make possible clear cases where you have 0.00% of gas in 1kpa mix.
This must make able clear pipes from ghost gases.

I make minimum mole to transfer to make able transfer all gas from gas container.
I try make machinery and pipe merge air like pipes do it.
Now it can be tested to how does this concept work, but it don't complite.

## Why It's Good For The Game

no more ghost near zero amounts of gases in airmixes.

## Changelog
:cl:
tweak: garbage_collect() now remove all gases with mole amount smaller than MINIMUM_MOLE_COUNT. No more ghost gases in airmixes.
/:cl:

### CODE REVIEW

**Comment by Dennok** [comment]
MINIMUM_MOLES_DELTA_TO_MOVE now 0.103984
i think need do something with this.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 1]
```diff
@@ -70,7 +70,7 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 /datum/gas_mixture/proc/garbage_collect(list/tocheck)
 	var/list/cached_gases = gases
 	for(var/id in (tocheck || cached_gases))
-		if(QUANTIZE(cached_gases[id][MOLES]) <= 0 && QUANTIZE(cached_gases[id][ARCHIVE]) <= 0)
+		if(QUANTIZE(cached_gases[id][MOLES]) < MINIMUM_MOLE_COUNT && QUANTIZE(cached_gases[id][ARCHIVE]) < MINIMUM_MOLE_COUNT)
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: I'm interested to see how this behaves ingame, because this might break a whole lot of things.
For context, let's look at the comment left on QUANTIZE so we can better understand the goal of this function.

`#define QUANTIZE(variable)		(round(variable,0.0000001))/*I feel the need to document what happens here. Basically this is used to catch most rounding errors, however it's previous value made it so that
															once gases got hot enough, most procedures wouldnt occur due to the fact that the mole counts would get rounded away. Thus, we lowered it a few orders of magnititude */`

So far as I know the use of quantize here is similar to changing the 0 to 0.0000001. You've raised this number to MINIMUM_MOLE_COUNT, which has a value of 0.01. 

If I understand the system correctly this will break fusion and kill gasses at high temperatures. I sympathize with the goal of removing those annoying 0kpa gasses from pipes but this is the wrong way to go about it.

Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Actually QUANTIZE used only in plasmafire
https://github.com/tgstation/tgstation/blob/51afca0dfc58bc50179b90db24c85b51b47601e4/code/modules/atmospherics/gasmixtures/reactions.dm#L179-L239
and garbage_collect()
https://github.com/tgstation/tgstation/blob/51afca0dfc58bc50179b90db24c85b51b47601e4/code/modules/atmospherics/gasmixtures/gas_mixture.dm#L70-L74


The MINIMUM_MOLE_COUNT maybe need make some smaller, from 0.01 to 0.001 or 0.0001, this need testing.

changing the 0 to 0.0000001 is some weird. This make imposible to delete gas from mix if it dont grabbed complitely.

Now i think about add some like MINIMUM_MOLE_MOVE = 0.01 or 0.001 to all places where moles moved partially, so no more pumps try move 0.00001 moles of gas infinitely.

Now i think about make atmos machinery work more similar to pipes. Now pipeline share gas from self pipe and all connected things by reconcile_air()
https://github.com/tgstation/tgstation/blob/51afca0dfc58bc50179b90db24c85b51b47601e4/code/modules/atmospherics/machinery/datum_pipeline.dm#L211-L264

I think about make machinery gases absorbed to pipeline so no more need reconcile_air() and machinery work directly to pipeline gas pool. 
As i see volume_pump already can use this type system
https://github.com/tgstation/tgstation/blob/51afca0dfc58bc50179b90db24c85b51b47601e4/code/modules/atmospherics/machinery/components/binary_devices/volume_pump.dm#L53-L86
 since it use `remove_ratio(transfer_rate/air1.volume)` where `air1.volume` = 200 now and `transfer_rate` max is 200.
When `air1.volume` merged in pipeline gas pull it still take only 200 liters from this big pipeline gas pull.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 2]
```diff
@@ -183,9 +189,15 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	for(var/new_id in gases_new)
+		cached_gases[new_id] = gases_new[new_id].Copy() //from ADD_GAS
+
 	//gas transfer
-	for(var/giver_id in giver_gases)
-		ASSERT_GAS(giver_id, src)
+	for(var/giver_id in gases_both)
 		cached_gases[giver_id][MOLES] += giver_gases[giver_id][MOLES]
 
 	return 1
```
Comment (Dennok) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: What think about this?

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 3]
```diff
@@ -183,9 +189,15 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	var/list/gases_new = giver_gases ^ gases_both
+
+	//add new gases
+	for(var/new_id in gases_new)
+		cached_gases[new_id] = gases_new[new_id].Copy() //from ADD_GAS
+
 	//gas transfer
-	for(var/giver_id in giver_gases)
-		ASSERT_GAS(giver_id, src)
+	for(var/giver_id in gases_both)
```
Comment (Dennok) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: What think about this?

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 4]
```diff
@@ -183,9 +189,15 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
 		cached_gases[giver_id][MOLES] += giver_gases[giver_id][MOLES]
```
Comment (Dennok) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: What think about this?

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 5]
```diff
@@ -183,9 +189,17 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	//add new gases
+	for(var/new_id in gases_new)
+		tmp_gas = gases_new[new_id]
+		cached_gases[new_id] = tmp_gas.Copy() //from ADD_GAS
+
 	//gas transfer
-	for(var/giver_id in giver_gases)
-		ASSERT_GAS(giver_id, src)
+	for(var/giver_id in gases_both)
 		cached_gases[giver_id][MOLES] += giver_gases[giver_id][MOLES]
```
Comment (Dennok) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: What think about this?

**Comment by CRITAWAKETS** [comment]
finally i dont have to care anymore about ghost gas

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 5 above)
Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: As I understand it merge() is called under the assumption that giver can be discarded afterwords.

The code you're replacing here is faster then or as fast as what you've replaced it with.
For one we have to spend time creating 3 new lists.
We then have to add all the gases together, not an expensive task. 

The painful part is the ^. As far as I can tell this means we have to remove all the gases that are in both and giver. So we are left with the gases in both that were not from giver, namely the gases from cashed.

This means that the for loop is just adding the gases that are in cashed back into cashed.

This is still the same or worse speed then ASSERT_GAS, which looks like this:
`if (!gas_mixture.gases[gas_id]) { ADD_GAS(gas_id, gas_mixture.gases) };`
That's 1 if check and an array access.
ADD_GAS looks like this:
`var/list/tmp_gaslist = GLOB.gaslist_cache[gas_id]; out_list[gas_id] = tmp_gaslist.Copy();`
Which is an array access and a copy().

Don't do this here.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 6]
```diff
@@ -5,8 +5,11 @@ What are the archived variables for?
 */
 #define MINIMUM_HEAT_CAPACITY	0.0003
 #define MINIMUM_MOLE_COUNT		0.01
-#define QUANTIZE(variable)		(round(variable,0.0000001))/*I feel the need to document what happens here. Basically this is used to catch most rounding errors, however it's previous value made it so that
+#define MINIMUM_MOLE_MANIPULATION_COUNT		0.001
+#define QUANTIZE(variable)		(round(variable, MINIMUM_MOLE_MANIPULATION_COUNT))/*I feel the need to document what happens here. Basically this is used to catch most rounding errors, however it's previous value made it so that
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Have you tested this? I am almost certain that this will break fusion and high temperature gas movement even more then it is now.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 7]
```diff
@@ -197,13 +220,15 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 	amount = min(amount, sum) //Can not take more air than tile has!
 	if(amount <= 0)
 		return null
+	var/ratio = amount / sum
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: This is a good damn change, thank you for spotting it.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 6 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: i test it now. Has problems with minimum mole transfer. it create negative moles that bypas garbage somewhere,
try place logging.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 8]
```diff
@@ -197,13 +220,15 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+	
 	var/datum/gas_mixture/removed = new type
 	var/list/removed_gases = removed.gases //accessing datum vars is slower than proc vars
 
 	removed.temperature = temperature
 	for(var/id in cached_gases)
 		ADD_GAS(id, removed.gases)
-		removed_gases[id][MOLES] = QUANTIZE((cached_gases[id][MOLES] / sum) * amount)
+		removed_gases[id][MOLES] = QUANTIZE_TO_MOVE((cached_gases[id][MOLES] * ratio)
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: What happens if you try and move something that has a molar count that's less then this? You'll just end up creating gas from nowhere right?

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 9]
```diff
@@ -221,7 +246,7 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 	removed.temperature = temperature
 	for(var/id in cached_gases)
 		ADD_GAS(id, removed.gases)
-		removed_gases[id][MOLES] = QUANTIZE(cached_gases[id][MOLES] * ratio)
+		removed_gases[id][MOLES] = QUANTIZE_TO_MOVE(cached_gases[id][MOLES] * ratio)
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: See above

Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: yes it now do something wrong. I think garbage delete negative micro moles but it pass it somewhere.

File: code/modules/atmospherics/machinery/datum_pipeline.dm [hunk 10]
```diff
@@ -90,11 +90,10 @@
...
 		var/list/adjacent = P.pipeline_expansion()
-		for(var/obj/machinery/atmospherics/pipe/I in adjacent)
-			if(I.parent == src)
-				continue
-			var/datum/pipeline/E = I.parent
-			merge(E)
+		for(var/obj/machinery/atmospherics/pipe/I in adjacent - members)
+			//if(I.parent == src)
+			//	continue
+			merge(I.parent)
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/machinery/datum_pipeline.dm]: Why did you change this?

Comment (Dennok) [reply code/modules/atmospherics/machinery/datum_pipeline.dm]: i think about remove all checks like  `f(I.parent == src)` 
i think it redundant

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 11]
```diff
@@ -169,7 +174,17 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
 	return 1
 
-/datum/gas_mixture/merge(datum/gas_mixture/giver)
+	//Merges all air and volume from giver into self
+	//Returns: 1 if we are mutable, 0 otherwise
+/datum/gas_mixture/proc/full_merge(datum/gas_mixture/giver)
+	if(!merge(giver))
+		return 0
+	volume += giver.volume
+	return 1
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: What is the intended use for this proc? Why would you want to merge volume?

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 6 above)
Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: If you can figure that out please let me know, I've been experiencing a similar issue.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 11 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: when i want really merge air conteyners. like pipelines.

Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Shouldn't pipenets do the work for you already? Or do you plan on giving those the rework they desperately need.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 8 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: yes, microquantities of gas, when gas near MINIMUM_MOLE_MANIPULATION_COUNT		0.001. 

In normal way it must be not called for zero mole gas, 
but it called now somewhere.

File: code/modules/atmospherics/machinery/datum_pipeline.dm (hunk 10 above)
Comment (LemonInTheDark) [reply code/modules/atmospherics/machinery/datum_pipeline.dm]: I don't think it is. The check should be there to prevent adding mols from yourself into yourself.

As far as I can tell you're attempting to fix that with the subtraction of members from adjacent, but that appears to be built to prevent merging with itself.

So it does what your subtraction does but without having to modify the list itself. I'm not sure if this is a performance benefit or downside, but you would need to profile it separately from all your changes here.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 8 above)
Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: I do not think allowing for gas to appear from nowhere is a good idea, it will either break a system or be abused by players.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 11 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: as i write upper
Now i think about make atmos machinery work more similar to pipes. Now pipeline share gas from self pipe and all connected things by reconcile_air()
i want remove reconcile_air() and make machinery work directly with pipeline air

Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: I'd recommend doing that in a separate pr so this one doesn't get slowed down by the idea.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 5 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: first i think replace part of ASSERT_GAS to ADD_GAS 
and next replace part with copy of global  and add moles to copy giver gas

>We then have to add all the gases together, not an expensive task.
wrong sentence, 
`gases_both = cached_gases & giver_gases` not summ. the result is a list that contains only items that were in both lists. so this is gases that need only mole change. no need assert.

next i remove this gases from list of new gases so i get list of gases that need ADD_GAS.

about creating of lists, what if put `cached_gases & giver_gases` directly in for, insteat of new var?

Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Ah I understand, forgive me.

Putting it in the for loop would help, I'm still not quite sure why you're using this over ASSERT_GAS.
Is there something I'm not getting?

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 8 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: this creating is unseable. If it noticeably something going wrong. Gas dont created from nowhere,
 to create air it must be at lower than mininun 0.001 moles and try take part of it.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 5 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: assert_gas use check, i think bypas checks can lower load. 
I dont wery sure in this change but if it remove some load why not implement it?

Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: I don't think your current method is cheaper then the assert_gas define. It goes about making the check for the gas in a different way but it does around the same thing. I can't say for sure because I don't know precisely how byond does ^ and & on lists.

It would be a good idea to profile it but you should probably do that on a clean master so it doesn't get effected by any of the other changes you've made here.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 11 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: now this is garbage of smal air ideas that i push here from work.
i clear it later or separate to new pr's. 
Now it place where someone who understands something here can say something about it.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 5 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: I think first better create clear byond test build to check this operations.

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 6 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: as i see now
/obj/item/organ/lungs/proc/check_breath() 
add zero gases to mix to use it in formulas
but `breath.garbage_collect()` on end of proc dont clear it propertly.

Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Do you have the info on the archived molar count when it's failing to gc?

Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: second number in check_breath() is ARCHIVE
```
		gasstring = ""
		for(var/gasid in breath.gases)
			gasstring += "[gasid] [breath.gases[gasid][MOLES]] [breath.gases[gasid][ARCHIVE]], "
		to_chat(world, "## TESTING: check_breath() b [gasstring]")
```
i test in https://github.com/Dennok/Dennok_tgstation/tree/gas-twixes build

Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: OMG, i screwed up
i remove `cached_gases -= id` from garbage_collect()
in the last check for sure

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 8 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: if we try take 50% of 0.001 moles of gas we move 0.001 moles. 
as i think if we have mole count lower than minimum outside of process (just lay on turf or in pipe or in ), something going wrong. If we have 0.003 moles and take 50% we take 0.0015 moles but QUANTIZE must change it to 0.002 moles, moles that we actual move.

Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: You madman. Maybe we should change it to an if. So if it's above the value we return it, and if not we return 0. I wonder....

Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: I'ma try that tomorrow. I wonder.....

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 5 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: From /datum/gas_mixture/proc/share(datum/gas_mixture/sharer, atmos_adjacent_turfs = 4)
https://github.com/tgstation/tgstation/blob/26d1edbe81f66ff52e8c4fd32bc55885dd4bc86a/code/modules/atmospherics/gasmixtures/gas_mixture.dm#L278-L281

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 8 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: Need test that `round(variable,0.001)` round 0.0005 to 0.001 and 0.0004 to 0

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 6 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: i fix garbage collect and now has right breath test
```
## TESTING: check_breath() Starting gas /datum/gas/oxygen 0.0044 0, /datum/gas/nitrogen 0.0164 0, /datum/gas/carbon_dioxide 0.001 0, 
## TESTING: check_breath() b /datum/gas/oxygen 0 0, /datum/gas/nitrogen 0.0164 0, /datum/gas/carbon_dioxide 0.0054 0, /datum/gas/plasma 0 0, /datum/gas/nitrous_oxide 0 0, /datum/gas/bz 0 0, /datum/gas/tritium 0 0, /datum/gas/nitryl 0 0, /datum/gas/pluoxium 0 0, /datum/gas/stimulum 0 0, /datum/gas/freon 0 0, 
## TESTING: garbage_collect(breath) 0 /datum/gas/oxygen 0 0
## TESTING: garbage_collect(breath) removed /datum/gas/oxygen
## TESTING: garbage_collect(breath) 0.0164 /datum/gas/nitrogen 0.0164 0
## TESTING: garbage_collect(breath) 0.0054 /datum/gas/carbon_dioxide 0.0054 0
## TESTING: garbage_collect(breath) 0 /datum/gas/plasma 0 0
## TESTING: garbage_collect(breath) removed /datum/gas/plasma
## TESTING: garbage_collect(breath) 0 /datum/gas/nitrous_oxide 0 0
## TESTING: garbage_collect(breath) removed /datum/gas/nitrous_oxide
## TESTING: garbage_collect(breath) 0 /datum/gas/bz 0 0
## TESTING: garbage_collect(breath) removed /datum/gas/bz
## TESTING: garbage_collect(breath) 0 /datum/gas/tritium 0 0
## TESTING: garbage_collect(breath) removed /datum/gas/tritium
## TESTING: garbage_collect(breath) 0 /datum/gas/nitryl 0 0
## TESTING: garbage_collect(breath) removed /datum/gas/nitryl
## TESTING: garbage_collect(breath) 0 /datum/gas/pluoxium 0 0
## TESTING: garbage_collect(breath) removed /datum/gas/pluoxium
## TESTING: garbage_collect(breath) 0 /datum/gas/stimulum 0 0
## TESTING: garbage_collect(breath) removed /datum/gas/stimulum
## TESTING: garbage_collect(breath) 0 /datum/gas/freon 0 0
## TESTING: garbage_collect(breath) removed /datum/gas/freon
## TESTING: check_breath() b /datum/gas/nitrogen 0.0164 0, /datum/gas/carbon_dioxide 0.0054 0, 
```

**Comment by Ghilker** [comment]
could this also fix the negative SM pressures and temp?

**Comment by LemonInTheDark** [comment]
I'll have a pr up tomorrow with the negative molar count fixes, that ok with you?

File: code/_compile_options.dm [hunk 12]
```diff
@@ -1,6 +1,7 @@
-//#define TESTING				//By using the testing("message") proc you can create debug-feedback for people with this
+#define TESTING				//By using the testing("message") proc you can create debug-feedback for people with this
 								//uncommented, but not visible in the release version)
-
+#define DEBUG
+#define LOWMEMORYMODE
```
Comment (duncathan) [inline code/_compile_options.dm]: ensure these are undone before any possible merge

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 13]
```diff
@@ -128,10 +135,18 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
...
+
 /datum/gas_mixture/proc/merge(datum/gas_mixture/giver)
 	if(!giver)
 		return 0
+	if(src == giver)//self merging is sin!
+		to_chat(world, "## TESTING: self merging is sin!")
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm]: is there a reason for this?

**Review by duncathan — CHANGES_REQUESTED** [review]
please remove the debugging messages. also can you please provide a more detailed explanation of these changes and their purpose? I don't quite understand what you're trying to do for much of this

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm (hunk 13 above)
Comment (Dennok) [reply code/modules/atmospherics/gasmixtures/gas_mixture.dm]: i remove `if(src == giver)` check  from places where `merge()` called. and need see that i dont mismach some logick and nothing called self merging. 
Selfmerging just increase gas amount twice.

**Comment by Dennok** [comment]
I make minimum mole to transfer to make able transfer all gas from gas container.
I try make machinery and pipe merge air like pipes do it.
Now it can be tested to how does this concept work, but it don't complite.
